session.go

```diff
@@ -31,182 +32,131 @@ const (
+	attachTimeout time.Duration
 	// command is the original command used to spawn the session.
 	command *Command
-	// cond broadcasts session changes.
+	// cond broadcasts session changes and any accompanying errors.
 	cond *sync.Cond
 	// configFile is the location of the screen configuration file.
 	configFile string
-	// error hold any error that occurred during a state change.
+	// error hold any error that occurred during a state change.  It is not safe
+	// to access outside of cond.L.
 	error error
 	// execer is used to spawn the session and ready commands.
 	execer Execer
-	// id holds the id of the session for both creating and attaching.
+	// id holds the id of the session for both creating and attaching.  This is
+	// generated uniquely for each session (rather than using the ID provided by
+	// the client) because without control of the daemon we do not have its PID
+	// and without the PID screen will do partial matching.  Enforcing a UUID
+	// should guarantee we match on the right session.
 	id string
+	// mutex prevents concurrent attaches to the session.  This is necessary since
+	// screen will happily spawn two separate sessions with the same name if
+	// multiple attaches happen in a close enough interval.  We are not able to
+	// control the daemon ourselves to prevent this because the daemon will spawn
+	// with a hardcoded 24x80 size which results in confusing padding above the
+	// prompt once the attach comes in and resizes.
+	mutex sync.Mutex
 	// options holds options for configuring the session.
 	options *Options
 	// socketsDir is the location of the directory where screen should put its
 	// sockets.
 	socketsDir string
-	// state holds the current session state.
+	// state holds the current session state.  It is not safe to access this
+	// outside of cond.L.
 	state State
-	// timer will close the session when it expires.
+	// timer will close the session when it expires.  The timer will be reset as
+	// long as there are active connections.
 	timer *time.Timer
 }
 
-// NewSession creates and immediately starts a new session.  Any errors with
-// starting are returned on Attach().  The session will close itself if nothing
-// is attached for the duration of the session timeout.
-func NewSession(id string, command *Command, execer Execer, options *Options) *Session {
+// NewSession sets up a new session.  Any errors with starting are returned on
+// Attach().  The session will close itself if nothing is attached for the
+// duration of the session timeout.
+func NewSession(command *Command, execer Execer, options *Options) *Session {
 	tempdir := filepath.Join(os.TempDir(), "coder-screen")
 	s := &Session{
-		command:    command,
-		cond:       sync.NewCond(&sync.Mutex{}),
-		configFile: filepath.Join(tempdir, "config"),
-		execer:     execer,
-		options:    options,
-		state:      StateStarting,
-		socketsDir: filepath.Join(tempdir, "sockets"),
+		attachTimeout: 30 * time.Second,
```

Comment:
This doesn't need to be a field on the session if we're just hardcoding the value.  Could make it a const.